### ISSUE TITLE
insertBatch to generate multiple INSERT statements

### ISSUE THREAD

**Issue by samdark** [yiisoft/db#888]
Right now, if the batch is too big, we're getting into limits of DBMS like the following:

![image](https://github.com/user-attachments/assets/207f92e4-d04b-4e00-97d7-78d4311daf53)

_The screenshot if from PostgreSQL._

I suggest doing two things:

1. Introduce configurable parameter to configure max number of parameters.
2. Default it per DBMS. For PostgreSQL it could be 65535.
3. If max is reached, move the rest of the data to the next `INSERT` statement.

**Comment by rob006** [comment]
That will change the characteristic of this operation: from atomic to non-atomic.

**Comment by samdark** [comment]
You can't make it atomically (except transactions) in case of exceeding max number of parameters anyway.

**Comment by rob006** [comment]
Yest, but implicit changing atomic transaction to non-atomic is just a massive footgun. This should be explicitly enabled by programmer who makes the call, so developer will know that this batch insert may result multiple queries.

Also, splitting one massive insert to multiple smaller ones is useful not only to handle parameters limit. I generally avoid really big inserts as they may block table for other queries, so system may be more responsive if you do 10 small inserts instead of one big one. Having a method that would do this for you would be useful even if I don't exceed parameters limit in your queries.

**Comment by samdark** [comment]
Makes sense. Default value of such option must be `null` then to disable splitting.

**Comment by Tigrov** [comment]
Good point, but
1. It is better if it works by default. Currently it does not work.
2. The documentation should mention this.

By default it is better to specify DBMS specific max value and allow the developer to change it.

**Comment by samdark** [comment]
Either way is fine for me, if we'll put a warning about it being atomic or not in the docs.

**Comment by rob006** [comment]
> By default it is better to specify driver specific max value and allow the developer to change it.

To be clear: I was talking about limit based on number of inserted rows, not number of params used by query. If I want to insert 50k records in 50 queries (1k records per query), then limit I need to pass to the function should `1000`.

**Comment by samdark** [comment]
Yes. That sounds useful as well.

**Comment by evil1** [comment]
Shouldn't this be done on DBMS level? In the following packages:
```
yiisoft/db-mssql
yiisoft/db-mysql
yiisoft/db-oracle
yiisoft/db-pgsql
yiisoft/db-sqlite
```

**Comment by evil1** [comment]
> > By default it is better to specify driver specific max value and allow the developer to change it.
> 
> To be clear: I was talking about limit based on number of inserted rows, not number of params used by query. If I want to insert 50k records in 50 queries (1k records per query), then limit I need to pass to the function should `1000`.

My vision is to introduce a parameter, let's say `batchInsertChunkSize` and default it to 0 (which means no limit), but if you set it to any non-negative value, then split insert queries in the appropriate chunks.

**Comment by samdark** [comment]
>  shouldn't this be done on DBMS level? 

Implementation itself — yes. The interface should be common, though.

**Comment by evil1** [comment]
> > shouldn't this be done on DBMS level?
> 
> Implementation itself — yes. The interface should be common, though.

Basically we are discussing here two similar questions at once.
As for the original issue described in the very first message - I think we can incapsulate this logic inside the DBMS pachage itself.
For PostgreSQL we could override the method `insertBatch()` in [DMLQueryBuilder](https://github.com/yiisoft/db-pgsql/blob/master/src/DMLQueryBuilder.php) in `yiisoft/db-pgsql` package, and split the insert batches into chunks, depending on the maximum number of params, which could be bound in PDO. Same for other packages.

**Comment by evil1** [comment]
This will require [insertBatch()](https://github.com/yiisoft/db/blob/4f9a12a01d93bcffe3e4fecc23c53ff86d70dd43/src/QueryBuilder/AbstractDMLQueryBuilder.php#L78) to return array of sql queries.

**Comment by evil1** [comment]
It seem's like there is no easy way, to implement these features.
I've reproduced the bug for pgsql
<img width="1799" alt="Снимок экрана 2024-12-27 в 16 09 25" src="https://github.com/user-attachments/assets/da459a38-afde-485c-aa59-aacb1b96e5e2" />

One of the ways is to create method `addRawSql()` and something like `addBindValues()`. And store `$sql` as an array. And iterate this queries on `execure()` call
```
    public function batchInsert(string $table, array $columns, iterable $rows): static
    {
        $table = $this->getQueryBuilder()->quoter()->quoteSql($table);

        /** @psalm-var string[] $columns */
        foreach ($columns as &$column) {
            $column = $this->getQueryBuilder()->quoter()->quoteSql($column);
        }

        unset($column);

        $params = [];
        $sql = $this->getQueryBuilder()->batchInsert($table, $columns, $rows, $params);

        $this->setRawSql($sql);
        $this->bindValues($params);

        return $this;
    }
```
